### PR TITLE
[Wget] Add project

### PR DESCRIPTION
**Pull request by rockdaboot**
While working on Debian unstable with clang-6.0, I see this error when running in the oss-fuzz docker image:
```
==11== ERROR: libFuzzer: fuzz target exited
    #0 0x4f90f3 in __sanitizer_print_stack_trace /src/llvm/projects/compiler-rt/lib/asan/asan_stack.cc:38
    #1 0x605124 in fuzzer::Fuzzer::ExitCallback() /src/libfuzzer/FuzzerLoop.cpp:248:5
    #2 0x7f0690fb3ff7  (/lib/x86_64-linux-gnu/libc.so.6+0x39ff7)
    #3 0x7f0690fb4044 in exit (/lib/x86_64-linux-gnu/libc.so.6+0x3a044)
    #4 0x539800 in main_wget /src/wget/src/main.c
    #5 0x5295b2 in LLVMFuzzerTestOneInput /src/wget/fuzz/wget_options_fuzzer.c:141:2
    #6 0x608c31 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/libfuzzer/FuzzerLoop.cpp:517:13
    #7 0x60b9f9 in fuzzer::Fuzzer::ReadAndExecuteSeedCorpora(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, fuzzer::fuzzer_allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&) /src/libfuzzer/FuzzerLoop.cpp:703:3
    #8 0x60c8f8 in fuzzer::Fuzzer::Loop(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, fuzzer::fuzzer_allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&) /src/libfuzzer/FuzzerLoop.cpp:741:3
    #9 0x5ec3fe in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/libfuzzer/FuzzerDriver.cpp:754:6
    #10 0x5df37c in main /src/libfuzzer/FuzzerMain.cpp:20:10
    #11 0x7f0690f9a82f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)
    #12 0x41e4c8 in _start (/out/wget_options_fuzzer+0x41e4c8)
```

The calls to exit() in the wget code base are catched using longjmp() and dlsym(). Does anything in the docker image prevent it ? Is there an 'official' way to get along with exit()s in fuzzed code ?

### CODE REVIEW

File: projects/wget/build.sh [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+# build and run non-networking tests
+LIBS="-lgnutls -lnettle -lhogweed -lidn2 -lunistring" \
+  ./configure -C
+make clean
+make -j$(nproc)
+make -j$(nproc) -C fuzz check
+
+# build for fuzzing
+LIBS="-lgnutls -lnettle -lhogweed -lidn2 -lunistring" \
+  ./configure --enable-fuzzing -C
```
Comment (inferno-chromium) [inline projects/wget/build.sh]: Does it need the other flags you use in wget2 ? --enable-static --disable-shared --disable-doc --without-plugin-support

File: projects/wget/build.sh [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+# avoid iconv() memleak on Ubuntu 16.04 image (breaks test suite)
+export ASAN_OPTIONS=detect_leaks=0
+
+cd $SRC/wget
+git checkout -b oss-fuzz origin/oss-fuzz
+./bootstrap
+
+# build and run non-networking tests
+LIBS="-lgnutls -lnettle -lhogweed -lidn2 -lunistring" \
+  ./configure -C
```
Comment (inferno-chromium) [inline projects/wget/build.sh]: Does it need the other flags you use in wget2 ?
--enable-static --disable-shared --disable-doc --without-plugin-support

File: projects/wget/build.sh [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    --disable-maintainer-mode --disable-libdane --disable-gcc-warnings --prefix=$WGET_DEPS_PATH $GNUTLS_CONFIGURE_FLAGS
+make -j$(nproc)
+make install
+
+
+# avoid iconv() memleak on Ubuntu 16.04 image (breaks test suite)
+export ASAN_OPTIONS=detect_leaks=0
+
+cd $SRC/wget
+git checkout -b oss-fuzz origin/oss-fuzz
```
Comment (inferno-chromium) [inline projects/wget/build.sh]: This seems wrong line here, can you remove.

File: projects/wget/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+ autoconf \
+ automake \
+ libtool \
+ texinfo \
+ flex \
+ bison \
+ gettext \
+ gengetopt \
+ curl \
+ gperf \
```
Comment (Dor1s) [inline projects/wget/Dockerfile]: are all these dependencies really needed?

**Comment by inferno-chromium** [comment]
@kcc @morehouse - any comment on the "fuzz target exited" question. in current scenario, i can add an exclusion to ignore crashes when we see that, but wondering if we should suggest something else.

File: projects/wget/build.sh (hunk 1 above)
Comment (rockdaboot) [reply projects/wget/build.sh]: No, Wget2 comes with a library, wget1.x doesn't.

File: projects/wget/build.sh (hunk 2 above)
Comment (rockdaboot) [reply projects/wget/build.sh]: No

File: projects/wget/build.sh (hunk 3 above)
Comment (rockdaboot) [reply projects/wget/build.sh]: Currently, the fuzzing code is in branch 'oss-fuzz' plus many small enhancements needed for fuzzing. Once all settles, I'll move it into master.

File: projects/wget/Dockerfile (hunk 4 above)
Comment (rockdaboot) [reply projects/wget/Dockerfile]: I fear yes, because of dependency building.

**Comment by morehouse** [comment]
Calls to `exit()` are considered crashes by libFuzzer.

@rockdaboot Are you sure exits should be considered OK in the wget case?

If you really want to ignore exits, you can have wget call `_exit()` instead.  This will not be caught by libFuzzer since `atexit` handlers will not be called.

**Comment by rockdaboot** [comment]
@morehouse  wget is a utility, calling exit() quite often. I catch all calls within wget to exit() and longjmp() back into LLVMFuzzerTestOneInput() where I cleanup and return 0. Calls to exit() outside LLVMFuzzerTestOneInput() use the original glibc function.

**Comment by morehouse** [comment]
AFAIK, we don't currently have a good way to disable libFuzzer's `atexit` hook.  We should probably make a flag for this.  I'll work on submitting a flag upstream in the next couple days.

**Pull request merged**